misk/src/main/kotlin/misk/eventrouter/RealEventRouter.kt

```diff
@@ -77,118 +86,171 @@ internal class RealEventRouter : EventRouter {
-          val socketEvent = SocketEvent.Event(action.topic, action.event.toString())
-          val eventJson = eventJsonAdapter.toJson(socketEvent)
-
-          if (topicOwner != clusterSnapshot.self) {
-            hostToSocket[topicOwner].send(eventJson)
-          } else {
-            remoteSubscribers.get(action.topic).forEach { it.send(eventJson) }
-            localSubscribers.get(action.topic).forEach { it.onEvent(socketEvent.message) }
-          }
+      is SocketEvent.Unsubscribe -> {
+        remoteSubscribers.get(socketEvent.topic).remove(action.webSocket)
+      }
+    }
+  }
+
+  private fun handleClusterChanged(action: Action.ClusterChanged) {
+    // TODO(tso): missed cluster changes? out of order cluster changes? figure
+    // out kubernetes watch behavior
+
+    logger.info { "handleClusterChanged: ${action.newSnapshot}" }
+
+    val topics = remoteSubscribers.keySet().plus(localSubscribers.keySet())
+    for (topic in topics) {
+      val localSubscribers = localSubscribers.get(topic)
+      val websockets = remoteSubscribers.get(topic)
+
+      if (clusterMapper.topicToHost(clusterSnapshot, topic) !=
+          clusterMapper.topicToHost(action.newSnapshot, topic)) {
+
+        val iterator = localSubscribers.iterator()
+        while (iterator.hasNext()) {
+          val localSubscription = iterator.next()
+          localSubscription.cancel()
+          iterator.remove()
         }
 
-        is Action.CancelSubscription -> {
-          val topicName = action.localSubscription.topic.name
-          val localTopicSubscribers = localSubscribers.get(topicName)
-          localTopicSubscribers.remove(action.localSubscription)
-          if (localTopicSubscribers.isEmpty()) {
-            val topicOwner = clusterMapper.topicToHost(clusterSnapshot, topicName)
-            if (topicOwner != clusterSnapshot.self) {
-              val unsubscribeEvent = eventJsonAdapter.toJson(SocketEvent.Unsubscribe(topicName))
-              hostToSocket[topicOwner].send(unsubscribeEvent)
-            }
-          }
-          action.localSubscription.onClose()
+        val wsIter = websockets.iterator()
+        while (wsIter.hasNext()) {
+          val next = wsIter.next()
+          next.close(1000, "the topic owner has changed")
+          wsIter.remove()
         }
       }
     }
+
+    clusterSnapshot = action.newSnapshot
+  }
+
+  private fun handleClosedWebSocket(action: Action.ClosedWebSocket) {
+    logger.info { "web socket closed: ${action.webSocket}" }
+
+    // TODO(tso): handle this more efficiently?
+    // this looks a lot like cluster changed. Maybe share code?
+    val hostname =
+        hostToSocket.asMap().entries.firstOrNull { it.value == action.webSocket }?.key ?: return
+    hostToSocket.invalidate(hostname)
+
+    val topics = localSubscribers.keySet()
+    for (topic in topics) {
+      val localSubscribers = localSubscribers.get(topic)
+
+      if (clusterMapper.topicToHost(clusterSnapshot, topic) == hostname) {
```

Comment:
I think this needs to check if the subscriber was subscribing via the closed socket, and not whether the hostnames match

Reply:
(do we ever have two websockets for a host?)

Reply:
... actually this is guarded by the code above on line 219. Cool.